src/opentracing-c/span.h

```diff
@@ -0,0 +1,252 @@
+
+    /**
+     * Adds a tag to the span. If there is a pre-existing tag set for `key`,
+     * it is overwritten. Tag values can be numeric types, strings, or bools.
+     * The behavior of other tag value types is undefined at the OpenTracing
+     * level. If a tracing system does not know how to handle a particular
+     * value type, it may ignore the tag, but shall not panic.
+     * @param span Span instance.
+     * @param key Tag key.
+     * @param value Tag value.
+     * @return opentracing_true if successful, opentracing_false otherwise
+     *         (indicates out of memory).
+     */
+    opentracing_bool (*set_tag)(struct opentracing_span* span,
+                                const char* key,
+                                const opentracing_value* value);
+
+    /**
+     * Record key:value logging data about a span.
+     * @param span Span instance.
+     * @param fields Array of log fields.
+     * @param num_fields Number of log fields.
+     * @return opentracing_true if successful, opentracing_false otherwise
+     *         (indicates out of memory).
+     * @see finish_with_options
+     */
+    opentracing_bool (*log_fields)(struct opentracing_span* span,
+                                   const opentracing_log_field* fields,
+                                   int num_fields);
+
+    /**
+     * Sets a key:value pair on this span and its span context that also
+     * propagates to descendants of this span. set_baggage_item() enables
+     * powerful functionality given a full-stack opentracing integration
+     * (e.g. arbitrary application data from a mobile app can make it,
+     * transparently, all the way into the depths of a storage system), and
+     * with it some powerful costs: use this feature with care.
+     * @attention set_baggage_item() will only propagate baggage items to
+     *            future causal descendants of the associated span.
+     * @attention Use this thoughtfully and with care. Every key and
+     *            value is copied into every local and remote child of the
+     *            associated span, and that can add up to a lot of network and
+     *            CPU overhead.
+     * @param span Span instance.
+     * @param key Baggage key.
+     * @param value Baggage value.
+     * @return opentracing_true if successful, opentracing_false otherwise
+     *         (indicates out of memory).
+     */
+    opentracing_bool (*set_baggage_item)(struct opentracing_span* span,
+                                         const char* key,
+                                         const char* value);
+
+    /**
+     * Gets the value for a baggage item given its key.
+     * @param span Span instance.
+     * @param key Baggage key.
+     * @return If baggage key found, returns baggage value, otherwise returns
+     *         empty string.
+     */
+    const char* (*baggage_item)(const struct opentracing_span* span,
+                                const char* key);
```

Comment:
Does this return an allocated string that must be freed?

If not, what happens if a user gets a baggage item, then calls set_baggage_item to set a new value for the key?

Reply:
That is a fair point, but I guess it relies heavily on the implementation of this map (which I have left undefined). My gut tells me this should be a case of invalidated pointer, similar to using getting an iterator from `std::vector`, pushing an item to the back of the vector, and attempting to dereference the initial iterator. There are no guarantees we won't need to reallocate the underlying buffer with that push, so all references are invalidated. If you are OK with that, I will add it to the docs.